File: app/src/androidTest/java/za/co/riggaroo/datecountdown/util/TaskExecutorWithIdlingResourceRule.java
package za.co.riggaroo.datecountdown.util;

import android.arch.core.executor.testing.CountingTaskExecutorRule;
import android.support.test.espresso.Espresso;
import android.support.test.espresso.IdlingResource;

import org.junit.runner.Description;

import java.util.concurrent.CopyOnWriteArrayList;

/**
 * A Junit rule that registers Architecture Components' background threads as an Espresso idling
 * resource.
 */
public class TaskExecutorWithIdlingResourceRule extends CountingTaskExecutorRule {
    private CopyOnWriteArrayList<IdlingResource.ResourceCallback> callbacks =
            new CopyOnWriteArrayList<>();
    @Override
    protected void starting(Description description) {
        Espresso.registerIdlingResources(new IdlingResource() {
            @Override
            public String getName() {
                return "architecture components idling resource";
            }

            @Override
            public boolean isIdleNow() {
                return TaskExecutorWithIdlingResourceRule.this.isIdle();
            }

            @Override
            public void registerIdleTransitionCallback(ResourceCallback callback) {
                callbacks.add(callback);
            }
        });
        super.starting(description);
    }

    @Override
    protected void onIdle() {
        super.onIdle();
        for (IdlingResource.ResourceCallback callback : callbacks) {
            callback.onTransitionToIdle();
        }
    }
}

File: app/src/androidTest/java/za/co/riggaroo/datecountdown/util/TaskExecutorWithIdlingResourceRule.kt
package za.co.riggaroo.datecountdown.util

import android.arch.core.executor.testing.CountingTaskExecutorRule
import android.support.test.espresso.Espresso
import android.support.test.espresso.IdlingResource

import org.junit.runner.Description

import java.util.concurrent.CopyOnWriteArrayList

/**
 * A Junit rule that registers Architecture Components' background threads as an Espresso idling
 * resource.
 */
class TaskExecutorWithIdlingResourceRule : CountingTaskExecutorRule() {
    private val callbacks = CopyOnWriteArrayList<IdlingResource.ResourceCallback>()
    override fun starting(description: Description?) {
        Espresso.registerIdlingResources(object : IdlingResource {
            override fun getName(): String {
                return "architecture components idling resource"
            }

            override fun isIdleNow(): Boolean {
                return this@TaskExecutorWithIdlingResourceRule.isIdle
            }

            override fun registerIdleTransitionCallback(callback: IdlingResource.ResourceCallback) {
                callbacks.add(callback)
            }
        })
        super.starting(description)
    }

    override fun onIdle() {
        super.onIdle()
        for (callback in callbacks) {
            callback.onTransitionToIdle()
        }
    }
}
